citest/gcp_testing/quota_predicate.py

```diff
@@ -0,0 +1,170 @@
+# Copyright 2016 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Support for observing and verifying GCP available quota."""
+
+import logging
+
+from . import GceContractBuilder
+
+from ..base import (
+  JournalLogger,
+  get_global_journal)
+
+from ..json_predicate import (
+  CompositePredicateResult,
+  DONT_ENUMERATE_TERMINAL,
+  NUM_GE,
+  EQUIVALENT,
+  PathPredicate,
+  PathValue,
+  PathValueResult,
+  ValuePredicate)
+
+
+class QuotaPredicate(ValuePredicate):
+  """Specialization to test GCP quota availability."""
+
+  @property
+  def minimum_quota(self):
+    """The expected quota"""
+    return self.__minimum_quota
+
+  def __init__(self, minimum_quota):
+    """Constructor.
+
+    Args:
+      minimum_quota [dict of number]: Dictionary of resource metric names
+         and quota they require. The predicate will check that the available
+         resource quota can accomodate the additional number in this dict.
+    """
+    self.__minimum_quota = dict(minimum_quota)
+    self.__logger = logging.getLogger(__name__)
+
+  def __repr__(self):
+    return 'minimum_metrics: {0!r}'.format(self.__minimum_quota)
+
+  def __call__(self, value):
+    """Implements ValuePredicate.
+
+    Args:
+       values: [array of dict] A list of dictionary resource quota descriptions
+          following the format returned described by the "quotas" attribute of
+          https://cloud.google.com/compute/docs/reference/latest/projects#resource
+    """
+    results = []
+    bad_metrics = []
+    available = {spec['metric']: spec['limit'] - spec['usage']
+                 for spec in value}
+    for metric, expect in self.__minimum_quota.items():
+      if not metric in available:
+        bad_metrics.append(metric)
+        self.__logger.info('Unknown metric "%s"', metric)
+
+        results.append(
+            PathValueResult(valid=False, pred=EQUIVALENT(metric),
+                            source=value, target_path='metric',
+                            path_value=PathValue('', available.keys())))
+        continue
+
+      found = available[metric]
+
+      # TODO(ewiseblatt): 20160715
+      # Need some way to express a DIFFERENCE of two paths to say
+      # that we want to evaluate ('limit' - 'usage') then report these values.
```

Comment:
Maybe this is too general, but I think something like a custom `LambdaPredicate` or something similar that takes a function and runs it on the input JSON would be useful. We could then extend that to basic +, -, *, / operations.